pallets/alt-vr/src/lib.rs

```diff
@@ -0,0 +1,203 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+#[cfg(test)]
+mod tests;
+
+mod default_weights;
+mod types;
+
+pub use default_weights::WeightInfo;
+pub use pallet::*;
+pub use types::*;
+
+use frame_support::traits::{Get, StorageVersion};
+
+use ternoa_primitives::ternoa;
+
+const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+
+    use frame_support::{ensure, pallet_prelude::*, transactional};
+
+    use frame_system::pallet_prelude::*;
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// Because this pallet emits events, it depends on the runtime's definition of an event.
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+        type WeightInfo: WeightInfo;
+
+        /// The minimum length a name string (username or vchatname) may be.
+        #[pallet::constant]
+        type MinNameLength: Get<u8>;
+
+        /// The maximum length a name string (username or vchatname) may be.
+        #[pallet::constant]
+        type MaxNameLength: Get<u8>;
+    }
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    #[pallet::storage_version(STORAGE_VERSION)]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Creates altvr informations.
+        #[pallet::weight(T::WeightInfo::create_altvr())]
+        #[transactional]
+        pub fn create_altvr(
+            origin: OriginFor<T>,
+            username: ternoa::String,
+            vchatname: ternoa::String,
+        ) -> DispatchResultWithPostInfo {
+            let owner = ensure_signed(origin)?;
+
+            let username_lower_bound = username.len() >= T::MinNameLength::get() as usize;
+            let username_upper_bound = username.len() <= T::MaxNameLength::get() as usize;
+            let vchatname_lower_bound = vchatname.len() >= T::MinNameLength::get() as usize;
+            let vchatname_upper_bound = vchatname.len() <= T::MaxNameLength::get() as usize;
+
+            ensure!(username_lower_bound, Error::<T>::TooShortUsername);
+            ensure!(username_upper_bound, Error::<T>::TooLongUsername);
+            ensure!(vchatname_lower_bound, Error::<T>::TooShortVchatname);
+            ensure!(vchatname_upper_bound, Error::<T>::TooLongVchatname);
+
+            // Create Altvr data
+            Self::new_altvr(&owner, &username, &vchatname);
```

Comment:
We should probably not allow user creation when that user already exists.

Reply:
done